### PR TITLE
Fix Issue #52505

### LINKED ISSUE #52505: BUG: incorrect reading of CSV  containing large integers
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
from io import StringIO

scsv = """#SYMBOL,SYSTEM,TYPE,MOMENT,ID,ACTION,PRICE,VOLUME,ID_DEAL,PRICE_DEAL
RIH3,F,S,20230301181139587,1925036343869802844,0,96690.00000,2,,
USDRUBF,F,S,20230301181139587,2023552585717888193,0,75.10000,14,,
USDRUBF,F,S,20230301181139587,2023552585717889863,1,75.00000,14,,
USDRUBF,F,S,20230301181139587,2023552585717889863,2,75.00000,1,2023552585717263358,75.00000
USDRUBF,F,B,20230301181139587,2023552585717882895,2,75.00000,1,2023552585717263358,75.00000
USDRUBF,F,S,20230301181139587,2023552585717889863,2,75.00000,1,2023552585717263359,75.00000
USDRUBF,F,B,20230301181139587,2023552585717888161,2,75.00000,1,2023552585717263359,75.00000
USDRUBF,F,S,20230301181139587,2023552585717889863,2,75.00000,10,2023552585717263360,75.00000
USDRUBF,F,B,20230301181139587,2023552585717889759,2,75.00000,10,2023552585717263360,75.00000
USDRUBF,F,S,20230301181139587,2023552585717889863,2,75.00000,2,2023552585717263361,75.00000
USDRUBF,F,B,20230301181139587,2023552585717889827,2,75.00000,2,2023552585717263361,75.00000
"""

for engine in "c pyarrow python".split():
    test = StringIO(scsv)
    orders = pd.read_csv(test, engine=engine,)
    print(engine, len(orders.query("ID_DEAL==2023552585717263360")))
```

### Issue Description

Reading of this financial data peace is terribly incorrect. I get the following output:

_c 8
pyarrow 8
python 0_

None of the engines is able to parse the data correctly, I suppose due to the integers being pretty big. But pandas produces no warning, just silently butchers the data. The issue is not new to 2.0, it was present in 1.5 as well.

It seems to be connected to float64 not having enough precision to hold all the digits. int64 would be ok probably, as the previous column works with int64, but current column has missing values.
using of the new nullable integer type kind of works, but not for pyarrow: 
...
    orders = pd.read_csv(test, engine=engine,dtype={"ID_DEAL": "Int64"})
...

_c 2
pyarrow 8
python 2_

But certainly the loss of precision needs to be recognized automatically, and either error should be issued, or Int64 dtype suggested. Otherwise, this bug can lead to catastrophical consequences in decision support systems powered by pandas.

### Expected Behavior

c 2
pyarrow 2
python 2

### Installed Versions

<details>
INSTALLED VERSIONS
------------------
commit           : 478d340667831908b5b4bf09a2787a11a14560c9
python           : 3.8.8.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.17763
machine          : AMD64
processor        : Intel64 Family 6 Model 45 Stepping 5, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : Russian_Russia.1251

</details>

### PR DESCRIPTION
**Pull request by manjalc**
github.com/pandas/issues/52505

- [X] closes #52505 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [X] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [X] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [X] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/io/parsers/readers.py [hunk 1]
```diff
@@ -587,6 +587,9 @@ def _read(
             raise ValueError(
                 "The 'chunksize' option is not supported with the 'pyarrow' engine"
             )
+        
+        if kwds.get("dtype_backend") is None or kwds.get("dtype_backend").value == 'NO_DEFAULT':
```
Comment (phofl) [inline pandas/io/parsers/readers.py:591]: What are you trying to do? This should not get set to pyarrow if not explicitly requested

**Review by phofl — CHANGES_REQUESTED** [review]

File: pandas/io/parsers/readers.py (hunk 1 above)
Comment (manjalc) [reply pandas/io/parsers/readers.py:591]: In #52505 when using the pyarrow engine and specifying the type_mapper, the read() function of ArrowParserWrapper (arrow_parser_wrapper.py) fails to create the pandas table with the appropriate type_mapper:

```Python
if self.kwds["dtype_backend"] == "pyarrow":
  frame = table.to_pandas(types_mapper=pd.ArrowDtype) # This is probably what should be returned
elif self.kwds["dtype_backend"] == "numpy_nullable":
  frame = table.to_pandas(types_mapper=_arrow_dtype_mapping().get)
else:
  frame = table.to_pandas() # This is being returned as dtype_backend is NO_DEFAULT, but this causes the values to be cast wrong
```

This commit sets the dtype_backend to "pyarrow" only if the engine is "pyarrow" and no other dtype_backend is specified so that the above code returns the proper data type. If there is a more appropriate way to approach this problem?

Comment (phofl) [reply pandas/io/parsers/readers.py:591]: Yeah and this is incorrect, numpy dtypes should be returned if no backend is set

Comment (manjalc) [reply pandas/io/parsers/readers.py:591]: Understood, my apologies. 

According to https://arrow.apache.org/docs/python/generated/pyarrow.Table.html#pyarrow.Table.to_pandas , you can pass dict.get directly to the type_mapper.

How about I submit a new pull request with the following change:

```Python
        # Line 163 in arrow_parser_wrapper.py
        if self.kwds["dtype_backend"] == "pyarrow":
            frame = table.to_pandas(types_mapper=pd.ArrowDtype)
        elif self.kwds["dtype_backend"] == "numpy_nullable":
            frame = table.to_pandas(types_mapper=_arrow_dtype_mapping().get)
        else:
            if self.kwds.get("dtype") is not None and isinstance(self.kwds.get("dtype"), dict):
                frame = table.to_pandas(types_mapper=self.kwds["dtype"].get)
            else:
                frame = table.to_pandas()
```

Would this change be more appropriate as it does not set the backend? I've checked that this also does fix the issue at hand (#52505)

Comment (phofl) [reply pandas/io/parsers/readers.py:591]: You can update this pr.

you could also handle the case where dtype is not none but also not a dict with this solution (potentially through a defaultdict)

Comment (manjalc) [reply pandas/io/parsers/readers.py]: Updated this pr. Changed the check to see if the dtype is a subclass of dict (which defaultdict is).

File: doc/source/whatsnew/v2.1.0.rst [hunk 2]
```diff
@@ -232,6 +232,7 @@ Numeric
 ^^^^^^^
 - Bug in :meth:`Series.corr` and :meth:`Series.cov` raising ``AttributeError`` for masked dtypes (:issue:`51422`)
 - Bug in :meth:`DataFrame.corrwith` raising ``NotImplementedError`` for pyarrow-backed dtypes (:issue:`52314`)
+- Bug in :func:`_read`, set dtype_backend to "pyarrow" if engine is "pyarrow" (:issue:`52505`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.0.rst:235]: Could you rephrase this describing the issue with `read_csv` from a user perspective?

File: pandas/io/parsers/arrow_parser_wrapper.py [hunk 3]
```diff
@@ -163,5 +163,8 @@ def read(self) -> DataFrame:
         elif self.kwds["dtype_backend"] == "numpy_nullable":
             frame = table.to_pandas(types_mapper=_arrow_dtype_mapping().get)
         else:
-            frame = table.to_pandas()
+            if self.kwds.get("dtype") is not None and issubclass(type(self.kwds.get("dtype")), dict):
+                frame = table.to_pandas(types_mapper=self.kwds["dtype"].get)
```
Comment (phofl) [inline pandas/io/parsers/arrow_parser_wrapper.py:167]: You could create a defaultdict when dtype is a scalar to cover this case as well.

But please add tests to check that this works

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.
